core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -293,60 +482,72 @@ public CreatePathResult createPath(AlluxioURI path, CreatePathOptions<?> options
         LOG.info("FileDoesNotExistException: {}", msg);
         throw new FileDoesNotExistException(msg);
-      } else {
-        // We will start filling at the index of the non-existing step found by the traversal.
-        pathIndex = traversalResult.getNonexistentPathIndex();
       }
     }
-
-    if (!traversalResult.getInode().isDirectory()) {
+    // The ancestor inode (parent or ancestor) of the target path.
+    Inode<?> ancestorInode = extensibleInodePath.getAncestorInode();
+    if (!ancestorInode.isDirectory()) {
       throw new InvalidPathException("Could not traverse to parent directory of path " + path
           + ". Component " + pathComponents[pathIndex - 1] + " is not a directory.");
     }
-    InodeDirectory currentInodeDirectory = (InodeDirectory) traversalResult.getInode();
+    InodeDirectory currentInodeDirectory = (InodeDirectory) ancestorInode;
+
     List<Inode<?>> createdInodes = Lists.newArrayList();
     List<Inode<?>> modifiedInodes = Lists.newArrayList();
-    // Directory persistence will not happen until the end of this method.
-    List<Inode<?>> toPersistDirectories = Lists.newArrayList(traversalResult.getNonPersisted());
-    if (pathIndex < parentPath.length || currentInodeDirectory.getChild(name) == null) {
+    // These are inodes that already exist, that should be journaled as persisted.
+    List<Inode<?>> existingNonPersisted = Lists.newArrayList();
+    // These are inodes to mark as persisted at the end of this method.
+    List<Inode<?>> toPersistDirectories = Lists.newArrayList();
+    if (options.isPersisted()) {
+      // Directory persistence will not happen until the end of this method.
+      toPersistDirectories.addAll(traversalResult.getNonPersisted());
+      existingNonPersisted.addAll(traversalResult.getNonPersisted());
+    }
+    if (pathIndex < (pathComponents.length - 1) || currentInodeDirectory.getChild(name) == null) {
       // (1) There are components in parent paths that need to be created. Or
       // (2) The last component of the path needs to be created.
       // In these two cases, the last traversed Inode will be modified.
       modifiedInodes.add(currentInodeDirectory);
     }
 
-    // Fill in the directories that were missing.
+    // Fill in the ancestor directories that were missing.
     CreateDirectoryOptions missingDirOptions = CreateDirectoryOptions.defaults()
         .setMountPoint(false)
         .setPersisted(options.isPersisted())
         .setPermissionStatus(options.getPermissionStatus());
-    for (int k = pathIndex; k < parentPath.length; k++) {
+    for (int k = pathIndex; k < (pathComponents.length - 1); k++) {
       InodeDirectory dir =
           InodeDirectory.create(mDirectoryIdGenerator.getNewDirectoryId(),
               currentInodeDirectory.getId(), pathComponents[k], missingDirOptions);
+      // Lock the newly created inode before subsequent operations, and add it to the lock group.
+      lockGroup.lockWrite(dir);
```

Comment:
presumably we are already holding a write lock on an ancestor, so there should be no need to hold this lock?